oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/UserPrincipalProviderCommitterProvider.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.jackrabbit.oak.security.user;
+
+import org.apache.jackrabbit.oak.api.Root;
+import org.apache.jackrabbit.oak.api.Tree;
+import org.jetbrains.annotations.NotNull;
+
+import java.security.Principal;
+import java.util.HashMap;
+import java.util.Set;
+
+public class UserPrincipalProviderCommitterProvider {
+
+    static UserPrincipalProviderCommitterProvider instance = null;
+    static HashMap<String, PrincipalCommitterThread> committerThreadMap = new HashMap<>();
```

Comment:
nitpicking: should be private static final as it only be used inside the UserPrincipalProviderCommitterProvider

on a more serious note: since the map is a singleton i believe it should deal with concurrent modifications as there might be concurrent logins for different users.

and another thing:
i am not super happy that the map is passed to the thread and the cleanup is done there. at a first glance i thought this was a bug because i didn't see the remove of the entry. so, if possible i would keep the cleanup in the provider.

Reply:
I modified it in a way that the provider has a method to remove elements from the map, and the thread call it.